source/core/slang-short-list.h

```diff
@@ -0,0 +1,487 @@
+        }
+
+        const T& getLast() const
+        {
+            SLANG_ASSERT(m_count > 0);
+            if (m_count <= shortListSize)
+                return m_shortBuffer[m_count - 1];
+            return m_buffer[m_count - shortListSize - 1];
+        }
+
+        T& getLast()
+        {
+            SLANG_ASSERT(m_count > 0);
+            if (m_count <= shortListSize)
+                return m_shortBuffer[m_count - 1];
+            return m_buffer[m_count - shortListSize - 1];
+        }
+
+        void removeLast()
+        {
+            SLANG_ASSERT(m_count > 0);
+            m_count--;
+        }
+
+        struct GetArrayViewResult
+        {
+            ArrayView<T> arrayView;
+            bool ownsStorage = false;
+
+            GetArrayViewResult() = default;
+            GetArrayViewResult(const GetArrayViewResult&) = delete;
+            GetArrayViewResult(GetArrayViewResult&& other)
+            {
+                ownsStorage = other.ownsStorage;
+                arrayView = other.arrayView;
+                other.ownsStorage = false;
+            }
+            ~GetArrayViewResult()
+            {
+                if (ownsStorage)
+                {
+                    ThisType::_free(arrayView.m_buffer, arrayView.m_count);
+                }
+            }
+        };
+        inline GetArrayViewResult getArrayView() const
+        {
+            GetArrayViewResult result;
+            if (m_count > shortListSize)
+            {
+                result.arrayView.m_buffer = ThisType::_allocate(m_count);
+                result.arrayView.m_count = m_count;
+                for (Index i = 0; i < shortListSize; i++)
+                    result.arrayView.m_buffer[i] = m_shortBuffer[i];
+                for (Index i = shortListSize; i < m_count; i++)
+                    result.arrayView.m_buffer[i] = m_buffer[i - shortListSize];
+                result.ownsStorage = true;
+            }
+            else
+            {
+                result.arrayView.m_buffer = const_cast<T*>(&m_shortBuffer[0]);
+                result.arrayView.m_count = m_count;
+            }
+            return result;
+        }
+
+        inline GetArrayViewResult getArrayView(Index start, Index count) const
+        {
+            SLANG_ASSERT(start >= 0 && count >= 0 && start + count <= m_count);
+            GetArrayViewResult result;
+            if (start < shortListSize && start + count > shortListSize)
+            {
+                result.ownsStorage = true;
+                result.arrayView.m_count = count;
+                result.arrayView.m_buffer = ThisType::_allocate(count);
+                for (Index i = start; i < shortListSize; i++)
+                    result.arrayView.m_buffer[i - start] = m_shortBuffer[i];
+                for (Index i = shortListSize; i < start + count; i++)
+                    result.arrayView.m_buffer[i - start] = m_buffer[i - shortListSize];
+                return result;
+            }
```

Comment:
Couldn't we simplify things by always moving all elements over to the allocated `m_buffer` as soon as you need more than `shortListSize` elements? It seems like we could streamline a lot of logic by having an invariant that things are always contiguous (e.g., it would also make the iterator a bit simpler - just a pointer).

Reply:
I thought about that and struggled between the two implementations. Maybe I made the wrong decision.

Reply:
I don't think we have to change it right away, but I do think that contiguous storage is an important property that we should only give up when there isn't a reasonable alternative.

One unrelated note is that sooner or later we will probably need/want to revisit the naming of the `List` type (which would affect `ShortList` too). Naming these as `Array<T>` and `ShortArray<T,N>` is probably the front-runner, but it would be good to reach consensus. The `List` name ends up confusing many people because most examples of "list" data structures are node-based rather than contiguous. Even something like Java `ArrayList` would be more clear about the underlying implementation.